cloudweatherreport/model.py

```diff
@@ -0,0 +1,764 @@
+        'date': datetime,
+        'test_id': basestring,
+        'provider_results': list([BenchmarkProviderResult]),
+    }
+
+    def provider_result(self, provider):
+        """
+        Returns the value of this benchmark result for the given provider.
+        """
+        for result in self.provider_results:
+            if result.provider == provider:
+                return result
+
+    def provider_value(self, provider):
+        """
+        Returns the value of this benchmark result for the given provider.
+        """
+        provider_result = self.provider_result(provider)
+        if provider_result:
+            return provider_result.value
+
+
+class Benchmark(BaseModel):
+    fields = {
+        'name': basestring,
+        'direction': basestring,
+        'units': basestring,
+        'results': list([BenchmarkResult]),
+    }
+
+    colors = [
+        '#0000ee',  # Blue
+        '#ff00ff',  # Magenta
+        '#009900',  # Green
+        '#ff8000',  # Orange
+        '#00cccc',  # Cyan
+        '#ee0000',  # Red
+        '#aa0099',  # Purple
+        '#006666',  # Dark Cyan
+        '#999999',  # Gray
+        '#dddd00',  # Yellow
+        '#000000',  # Black
+    ]
+
+    @classmethod
+    def from_action(cls, action_result):
+        try:
+            # actions don't preserve type, so coerce to float
+            value = float(action_result['value'])
+        except ValueError as e:
+            log.warn('Skipping malformed benchmark value for %s: %s',
+                     action_result['name'], e.message)
+            value = None
+        return cls(
+            name=action_result['name'],
+            direction=action_result['direction'],
+            units=action_result['units'],
+            results=[
+                BenchmarkResult(
+                    test_id=action_result['test_id'],
+                    date=datetime.now(),
+                    provider_results=[
+                        BenchmarkProviderResult(
+                            provider=action_result['provider'],
+                            value=value,
+                        ),
+                    ],
+                ),
+            ],
+        )
+
+    @property
+    def providers(self):
+        return sorted({pr.provider
+                       for r in self.results
+                       for pr in r.provider_results})
+
+    def result_by_test_id(self, test_id, create=False):
+        for result in self.results:
+            if result.test_id == test_id:
+                return result
+        if create:
+            result = BenchmarkResult(
+                test_id=test_id,
+                date=datetime.now(),
+            )
+            self.results.append(result)
+            return result
+
+    def as_chart(self):
+        """
+        Serialize this Benchmark to chart data as a Python dict.
+
+        The result is used to render a graph using Chart.js.
+        """
+        data = {
+            'title': self.name,
```

Comment:
The chart data don't make much sense unless we include `units` and `direction` info.  I think they should be included something like:
`Terasort Benchmark Chart (Units: secs  Direction:asc)`